az://303 vclock enhancements redux

jonmeredith opened the pull request:
Final version of the node-based vclock work.  

The put FSM has been changed in the following ways
- Puts must run on a node in the preference list.  If not in the preflist it will be forwarded by the put FSM.
- Puts are written to the local vnode before being written to the remotes.
- Vclocks are no longer incremented by riak_client, they are done by the first vnode in the preference list. 
- The Id for the vector clock update is based on the node name and start time for the vnode.
- The changes should significantly reduce the number of entries present in the vclocks and make pruning a very unusual event.
- Because the vclock updates are now guaranteed to create vclocks that are frontiers the introduction of extra vclock ids in riak_kv_vnode:syntactic_put_merge has now been removed, getting rid of a class of vclock explosion problems.

This p/r replaces the original for az://303 and should be reviewed with an eye to the comments raised by Bryan/Scott being addressed.  It needs to be reviewed in conjunction with the riak_core az://303 pull request.

Commit:
Eventual consistency test.

Currently just a framework for writing tests modelling clients and
processes.

Commit:
Added get to ec_eqc.

Commit:
Added skeleton put request to ec_eqc.

Commit:
Added priorities to messages - now scrambles delivery order.

Commit:
Refactored request into a record, removed pri from client.

Commit:
Fixed comments.

Commit:
Changed fsm model in ec_eqc to use get/put core.

Commit:
Ensured last request is a get.

Commit:
Check expected values in results.

Found first counterexample.

Failed!
{[0,0,1,0],
 [{client,undefined,[update,get],undefined},
  {client,undefined,[update],undefined}],
 {params,undefined,0,1,1,undefined}}
Params:
{params,undefined,1,1,1,1}
Clients:
[{client,1,
         [{req,0,undefined,{get,[{kv_vnode,1,1}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1}]}},
          {req,1,undefined,{get,[{kv_vnode,1,1}]}}],
         undefined},
 {client,2,
         [{req,0,undefined,{get,[{kv_vnode,1,1}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1}]}}],
         undefined},
 {client,1000001,[{req,1000002,undefined,{get,[{kv_vnode,1,1}]}}],undefined}]
History:
[{deliver_msg,{msg,0,{req,2},{get_fsm,2},{get,[{kv_vnode,1,1}]}}},
 {deliver_msg,{msg,0,{get_fsm,2},{kv_vnode,1,1},{get,2}}},
 {deliver_msg,{msg,0,{kv_vnode,1,1},{get_fsm,2},{r,{error,notfound},1,2}}},
 {deliver_msg,{msg,1,{get_fsm,2},{req,2},{error,notfound}}},
 {result,2,{req,0,2,{get,[{kv_vnode,1,1}]}},{error,notfound}},
 {deliver_msg,
     {msg,0,
         {req,3},
         {put_fsm,3},
         {put,
             [{kv_vnode,1,1}],
             {r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,3>>}],
                 [{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {deliver_msg,
     {msg,0,
         {put_fsm,3},
         {kv_vnode,1,1},
         {put,
             {r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,3>>}],
                 [{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined},
             3,3}}},
 {deliver_msg,{msg,0,{kv_vnode,1,1},{put_fsm,3},{w,1,3}}},
 {deliver_msg,{msg,1,{req,1},{get_fsm,1},{get,[{kv_vnode,1,1}]}}},
 {deliver_msg,{msg,0,{get_fsm,1},{kv_vnode,1,1},{get,1}}},
 {deliver_msg,
     {msg,0,
         {kv_vnode,1,1},
         {get_fsm,1},
         {r,{ok,{r_object,<<"b">>,<<"k">>,
                    [{r_content,
                         {dict,0,16,16,8,80,48,
                             {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                             {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                               []}}},
                         <<0,0,0,3>>}],
                    [{<<0,0,0,2>>,{1,63472389426}}],
                    {dict,1,16,16,8,80,48,
                        {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                        {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                          [[clean|true]],
                          []}}},
                    undefined}},
            1,1}}},
 {deliver_msg,
     {msg,0,
         {get_fsm,1},
         {req,1},
         {ok,{r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,3>>}],
                 [{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {result,1,
     {req,0,1,{get,[{kv_vnode,1,1}]}},
     {ok,{r_object,<<"b">>,<<"k">>,
             [{r_content,
                  {dict,0,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                  <<0,0,0,3>>}],
             [{<<0,0,0,2>>,{1,63472389426}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}}},
 {deliver_msg,
     {msg,1,
         {req,4},
         {put_fsm,4},
         {put,
             [{kv_vnode,1,1}],
             {r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,4>>}],
                 [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {deliver_msg,
     {msg,0,
         {put_fsm,4},
         {kv_vnode,1,1},
         {put,
             {r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,4>>}],
                 [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined},
             4,4}}},
 {deliver_msg,{msg,0,{kv_vnode,1,1},{put_fsm,4},{dw,1,4}}},
 {deliver_msg,{msg,0,{kv_vnode,1,1},{put_fsm,4},{w,1,4}}},
 {deliver_msg,{msg,1,{put_fsm,4},{req,4},ok}},
 {result,1,{req,0,4,{put,[{kv_vnode,1,1}]}},ok},
 {deliver_msg,{msg,1,{kv_vnode,1,1},{put_fsm,3},{dw,1,3}}},
 {deliver_msg,{msg,0,{put_fsm,3},{req,3},ok}},
 {result,2,{req,0,3,{put,[{kv_vnode,1,1}]}},ok},
 {deliver_msg,{msg,0,{req,5},{get_fsm,5},{get,[{kv_vnode,1,1}]}}},
 {deliver_msg,{msg,0,{get_fsm,5},{kv_vnode,1,1},{get,5}}},
 {deliver_msg,
     {msg,1,
         {kv_vnode,1,1},
         {get_fsm,5},
         {r,{ok,{r_object,<<"b">>,<<"k">>,
                    [{r_content,
                         {dict,0,16,16,8,80,48,
                             {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                             {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                               []}}},
                         <<0,0,0,4>>}],
                    [{<<0,0,0,1>>,{1,63472389426}},
                     {<<0,0,0,2>>,{1,63472389426}}],
                    {dict,1,16,16,8,80,48,
                        {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                        {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                          [[clean|true]],
                          []}}},
                    undefined}},
            1,5}}},
 {deliver_msg,
     {msg,0,
         {get_fsm,5},
         {req,5},
         {ok,{r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,4>>}],
                 [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {result,1,
     {req,1,5,{get,[{kv_vnode,1,1}]}},
     {ok,{r_object,<<"b">>,<<"k">>,
             [{r_content,
                  {dict,0,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                  <<0,0,0,4>>}],
             [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}}},
 {deliver_msg,{msg,0,{req,6},{get_fsm,6},{get,[{kv_vnode,1,1}]}}},
 {deliver_msg,{msg,0,{get_fsm,6},{kv_vnode,1,1},{get,6}}},
 {deliver_msg,
     {msg,1,
         {kv_vnode,1,1},
         {get_fsm,6},
         {r,{ok,{r_object,<<"b">>,<<"k">>,
                    [{r_content,
                         {dict,0,16,16,8,80,48,
                             {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                             {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                               []}}},
                         <<0,0,0,4>>}],
                    [{<<0,0,0,1>>,{1,63472389426}},
                     {<<0,0,0,2>>,{1,63472389426}}],
                    {dict,1,16,16,8,80,48,
                        {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                        {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                          [[clean|true]],
                          []}}},
                    undefined}},
            1,6}}},
 {deliver_msg,
     {msg,0,
         {get_fsm,6},
         {req,6},
         {ok,{r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,4>>}],
                 [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {result,1000001,
     {req,1000002,6,{get,[{kv_vnode,1,1}]}},
     {ok,{r_object,<<"b">>,<<"k">>,
             [{r_content,
                  {dict,0,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                  <<0,0,0,4>>}],
             [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}}}]
must: failed
[<<0,0,0,3>>] /= []
false

Commit:
Fixed model for counterexample - needed to add checks around out-of-order put results.

XXX Cid 2 GOT VALUES: []
XXX Cid 1 GOT VALUES: [<<0,0,0,3>>]
XXX Cid 1 PUT <<0,0,0,4>> OVER VALUES: [<<0,0,0,3>>]
XXX Cid 2 PUT <<0,0,0,3>> OVER VALUES: []
XXX Cid 1 GOT VALUES: [<<0,0,0,4>>]

The second put now checks the value is no longer in may.

Commit:
Another fix for out of order put responses.

Counterexample

[{[0,0,2,0,2,1,3],
  [{client,undefined,[get],undefined},
   {client,undefined,[update,get,update],undefined},
   {client,undefined,[get],undefined},
   {client,undefined,[update,get],undefined}],
  {params,undefined,0,2,2,undefined}}]

6> eqc:check(ec_eqc:prop(), eqc:current_counterexample()).
XXX Cid 2 GOT VALUES: []
XXX Cid 4 GOT VALUES: [<<0,0,0,4>>]
XXX Cid 2 PUT <<0,0,0,4>> OVER VALUES: []  MUST: [<<0,0,0,4>>] MAY: []
XXX Cid 2 GOT VALUES: [<<0,0,0,5>>]
XXX Cid 2 GOT VALUES: [<<0,0,0,5>>]
XXX Cid 2 PUT <<0,0,0,8>> OVER VALUES: [<<0,0,0,5>>]  MUST: [<<0,0,0,8>>,
                                                             <<0,0,0,4>>] MAY: [<<0,0,0,5>>]
XXX Cid 1 GOT VALUES: []
XXX Cid 4 PUT <<0,0,0,5>> OVER VALUES: [<<0,0,0,4>>]  MUST: [<<0,0,0,8>>] MAY: [<<0,0,0,5>>]
XXX Cid 4 GOT VALUES: [<<0,0,0,8>>]
XXX Cid 3 GOT VALUES: [<<0,0,0,8>>]
XXX Final GOT VALUES: [<<0,0,0,8>>] Must: [<<0,0,0,8>>] May: [<<0,0,0,5>>]
OK, passed the test.

Commit:
EC test passes for preflist made of primary nodes.

Commit:
Added handoff capability/final handoff but still using primaries in preflist.

Commit:
Added fallback nodes to preflists - get expected failures.

C1/C2 both update the value once on the primaries and C2 also updates
on fallback nodes only.  Because the primary VC is [{C1,1},{C2,1}] the
object on the fallback nodes is ignored with VC [{C2,1}]

61> ec_eqc:check().
XXXCid 1 GOT VALUES: []
XXXCid 1 PUT <<0,0,0,2>> OVER VALUES: []  MUST: [<<0,0,0,2>>] MAY: []
XXXCid 2 GOT VALUES: [<<0,0,0,2>>]
XXXCid 2 PUT <<0,0,0,4>> OVER VALUES: [<<0,0,0,2>>]  MUST: [<<0,0,0,4>>] MAY: [<<0,0,0,2>>]
XXXCid 2 GOT VALUES: []
XXXCid 2 PUT <<0,0,0,6>> OVER VALUES: []  MUST: [<<0,0,0,6>>,<<0,0,0,4>>] MAY: [<<0,0,0,2>>]
XXXFinal GOT VALUES: [<<0,0,0,4>>] Must: [<<0,0,0,6>>,<<0,0,0,4>>] May: [<<0,0,0,2>>]
Failed!
{[0],
 [{client,undefined,[{update,[0],[0]}],undefined},
  {client,undefined,[{update,[0],[0]},{update,[5],[0]}],undefined}],
 {params,undefined,0,1,2,undefined}}
Params:
{params,[{m,2},{n,2},{r,1},{w,2},{dw,2}]}
Clients:
[{client,1,
         [{req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined},
 {client,2,
         [{req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{get,[{kv_vnode,1,2},{kv_vnode,2,1}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined},
 {client,1000000,
         [{req,1000001,undefined,handoff_fallbacks},
          {req,1000002,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined}]

Commit:
Changed to only allowing one fallback per preflist, still has problems.

70> ec_eqc:check().
XXXCid 1 GOT VALUES: []
XXXCid 1 PUT <<0,0,0,2>> OVER VALUES: []  MUST: [<<0,0,0,2>>] MAY: []
XXXCid 2 GOT VALUES: []
XXXCid 2 PUT <<0,0,0,4>> OVER VALUES: []  MUST: [<<0,0,0,4>>,<<0,0,0,2>>] MAY: []
XXXCid 2 GOT VALUES: [<<0,0,0,4>>]
XXXCid 2 PUT <<0,0,0,6>> OVER VALUES: [<<0,0,0,4>>]  MUST: [<<0,0,0,6>>,
                                                            <<0,0,0,2>>] MAY: [<<0,0,0,4>>]
XXXCid 2 GOT VALUES: [<<0,0,0,6>>]
XXXCid 2 PUT <<0,0,0,8>> OVER VALUES: [<<0,0,0,6>>]  MUST: [<<0,0,0,8>>,
                                                            <<0,0,0,2>>] MAY: [<<0,0,0,4>>,
                                                                               <<0,0,0,6>>]
XXXCid 2 GOT VALUES: [<<0,0,0,6>>]
XXXCid 2 PUT <<0,0,0,10>> OVER VALUES: [<<0,0,0,6>>]  MUST: [<<0,0,0,10>>,
                                                             <<0,0,0,2>>,
                                                             <<0,0,0,8>>] MAY: [<<0,0,0,4>>,
                                                                                <<0,0,0,6>>]
XXXFinal GOT VALUES: [<<0,0,0,2>>,<<0,0,0,8>>] Must: [<<0,0,0,10>>,
                                                      <<0,0,0,2>>,
                                                      <<0,0,0,8>>] May: [<<0,0,0,4>>,
                                                                         <<0,0,0,6>>]
Failed!
{[0],
 [{client,undefined,[{update,[0],[2,5]}],undefined},
  {client,undefined,
          [{update,[0],[0]},
           {update,[0],[0]},
           {update,[0],[2,11]},
           {update,[0],[0]}],
          undefined}],
 {params,undefined,0,1,2,undefined}}
Params:
{params,[{m,2},{n,2},{r,1},{w,2},{dw,2}]}
Clients:
[{client,1,
         [{req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,2},{kv_vnode,2,2}]}}],
         undefined},
 {client,2,
         [{req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,2},{kv_vnode,2,2}]}},
          {req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined},
 {client,1000000,
         [{req,1000001,undefined,handoff_fallbacks},
          {req,1000002,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined}]

Commit:
Even a single fallback causes problems.

114> ec_eqc:check().
XXXCid 1 GOT VALUES: [] FALLBACKS: []
XXXCid 1 PUT <<"C1-U1">> OVER [] MUST: [<<"C1-U1">>] MAY: [] FALLBACKS: []
XXXCid 1 GOT VALUES: [] FALLBACKS: [{1,2}]
XXXCid 1 PUT <<"C1-U2">> OVER [] MUST: [<<"C1-U2">>,<<"C1-U1">>] MAY: [] FALLBACKS: [{1,
                                                                                      2}]
XXXCid 1 GOT VALUES: [<<"C1-U1">>] FALLBACKS: []
XXXCid 1 PUT <<"C1-U3">> OVER [<<"C1-U1">>] MUST: [<<"C1-U3">>,<<"C1-U2">>] MAY: [<<"C1-U1">>] FALLBACKS: []
XXXFinal GOT VALUES: [<<"C1-U3">>] Must: [<<"C1-U3">>,<<"C1-U2">>] May: [<<"C1-U1">>] Fallbacks: []
Failed!
{[0],
 [{client,undefined,
          [{update,[0],[0]},{update,[6,1],[6,1]},{update,[0],[0]}],
          undefined}],
 {params,undefined,0,1,3,undefined}}
Params:
{params,[{m,3},{n,3},{r,1},{w,3},{dw,3}]}
Clients:
[{client,1,
         [{req,0,undefined,
               {get,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}]}},
          {req,0,undefined,
               {put,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}],
                    <<"C1-U1">>}},
          {req,0,undefined,
               {get,[{kv_vnode,1,2},{kv_vnode,2,2},{kv_vnode,3,3}]}},
          {req,0,undefined,
               {put,[{kv_vnode,1,2},{kv_vnode,2,2},{kv_vnode,3,3}],
                    <<"C1-U2">>}},
          {req,0,undefined,
               {get,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}]}},
          {req,0,undefined,
               {put,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}],
                    <<"C1-U3">>}}],
         undefined},
 {client,1000000,
         [{req,1000001,undefined,handoff_fallbacks},
          {req,1000002,undefined,
               {get,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}]}}],
         undefined}]

Commit:
Tried generating fail message on failed-due-to-out-of-date-vclock writes

Reverted to generating preference lists using all nodes.
Still fails.

149> ec_eqc:check().
XXXCid 1 GOT VALUES: [] FALLBACKS: []
XXXCid 1 PUT <<"C1-U1">> OVER [] MUST: [] MAY: [] FALLBACKS: [{1,2},
                                                              {2,3},
                                                              {3,1}]
XXXCid 2 GOT VALUES: [] FALLBACKS: []
XXXCid 2 PUT <<"C2-U1">> OVER [] MUST: [] MAY: [] FALLBACKS: [{1,2},
                                                              {2,3},
                                                              {3,1}]
XXXCid 2 GOT VALUES: [] FALLBACKS: []
XXXCid 2 PUT <<"C2-U2">> OVER [] MUST: [<<"C2-U2">>] MAY: [] FALLBACKS: []
XXXFinal GOT VALUES: [<<"C1-U1">>,<<"C2-U1">>] Must: [<<"C2-U2">>] May: [] Fallbacks: []
Failed!
{[0],
 [{client,undefined,[{update,[0],[16]}],undefined},
  {client,undefined,[{update,[0],[1]},{update,[0],[0]}],undefined}],
 {params,undefined,0,1,2,undefined}}
Params:
{params,[{m,3},{n,3},{r,2},{w,2},{dw,2}]}

Commit:
Still getting inconsistencies despite writing to W=all, DW=all when there are fallbacks.

Commit:
Changed put FSM to do dynamo-style coordinator vnode put first then remote.

Commit:
Reverted to generating parameters again - put still has W/DW=all

Commit:
Make w/dw generated as part of the test.

Commit:
Added read repair.

Commit:
Got handoff working with object going immediately undefined.

In the real code the handoff transmission/deletion is async.
Handoff starts, network connection gets set up, a fold kicks off
and once all data is transmitted the vnode does the drop.  It could
service other requests in the mean time.  In fact it looks
like the data is dropped on handoff completion, even if something
has arrived in the mean time.

Commit:
Made handoff/cleanup async.

Commit:
Changed coordinating vnode to increment clock if necessary.

Commit:
Changed put FSM / kv vnode to support dynamo style puts.

Put FSM test currently broken.

Commit:
Fixed missing state update and incorrectly accessing vector clock counters.

Commit:
Updated put FSM eqc tests for vclock changes.

Commit:
Changed ec_eqc to handoff to random partition, not just owner.

Used to demonstrate issue spotted by Bryan,
https://github.com/basho/riak_kv/pull/127#issuecomment-1554900
This is equivalent to the ownership change described in the comment
and lets us test sequences like this:

dev2: create object
dev2: update object (object now [{dev2,2}],v1b
dev2: reassign partition
dev2: wait for transfer to complete
dev1: get object to make sure it's there
dev1: shutdown
dev2: get object, check notfound
dev2: create object (vclock now [{dev2,1],v1)
dev2: check object
dev1: startup, wait for handoff to complete.
dev1: sure enough, v1b is the object that is found.

But without introducing owners to the model.  I don't think it's necessary
as the final handoff provides the synchronization point needed to check
eventual consistency.  Before that point any vnode should handoff to any
other.

Commit:
Stopped ec_eqc test from deleting primary partitions after transfer.

Commit:
Prepared ec_eqc datastructures for modelling ownership.

Commit:
ec_eqc test now demonstrates issues around dropping vnodes.

Commit:
Updated ec_eqc and put fsm/vnode to add vnode epoch to vclock node id.

Fixes the case identified by Bryan in the pull request comments, where
a vnode can transfer ownership (deleting the local copy of the data) then
be used as a partitioned fallback.  Previously that would have re-used
a vclock entry, but now the vnode epoch is included in the clock
to guarantee uniqueness.

Commit:
Cleaned up the code to create a per-node/vnode vclock id.

Also modified vclock:get_counter to behave the way the wrapper
was working - no other consumers were using it.

Commit:
Updated put FSM eqc mock vnode tests to do match coord puts in kv_vnode.

Commit:
Extended put FSM eqc test to ensure correct number of put requests received.

Commit:
Fixed case where put requests not sent to vnodes W/DW=1.

Required changes to the put EQC model to work out how many
responses should be expected and added a special case for
local vnode failure.

Commit:
Added check for riak_kv:vnode_vclocks app var to enable new behavior.

When running with it set to anything but true it will increment
by the client id before doing the local put - this is only really
needed for forwarding to other nodes if not in the preference list
as the local node will support a vnode coord put.

jonmeredith commented:
riak_core pull request: https://github.com/basho/riak_core/pull/72
original pull request: https://github.com/basho/riak_kv/pull/127

argv0 commented:
i didn't do a thorough review, but I've done a significant amount of local testing with this PR and saw no errors.

beerriot commented on src/riak_kv_put_core.erl:
```diff
@@ -90,11 +95,17 @@ add_result(_Other, PutCore = #putcore{num_fail = NumFail}) ->
+%% True if coordinator result has been returned - anything other than w.
+coord_result(_CoordIdx, _Results) ->
+    true.
+%    [Result || {Idx, Result} <- Results, Idx == CoordIdx] -- [w] /= [].
```
This looks like a debugging modification that was left in - is it?

jonmeredith replied:
Yes, was an early stab at an alternative method where the put request got sent to all nodes with one marked as a coordinator and if you discovered you weren't creating a frontier vclock then you resubmitted.  That method may still work but I hit some problems with it and decided to simplify.

beerriot commented on src/riak_kv_put_fsm.erl:
```diff
@@ -71,6 +75,7 @@
                 n :: pos_integer(),
                 w :: non_neg_integer(),
                 dw :: non_neg_integer(),
+                coord_idx_node,
```
`:: atom()`?

jonmeredith replied:
:: {integer(), atom()} - it's actually the coordinator preflist entry so I've renamed it to coord_pl_entry instead

beerriot commented on src/riak_kv_put_fsm.erl:
```diff
@@ -258,35 +289,51 @@ precommit(timeout, State = #state{precommit = [Hook | Rest], robj = RObj}) ->
+                                         coord_idx_node = {_Index, _Node} = CoordIndexNode,
+                                         vnode_options=VnodeOptions,
+                                         starttime = StartTime}) ->
+    %RObj = riak_object:increment_vclock(RObj0, Node, StartTime),
```
`<General call for not leaving unused code hanging around commented />`

jonmeredith replied:
ty - fixed.

beerriot commented on src/riak_kv_stat.erl:
```diff
@@ -410,7 +414,8 @@ node_stats(Moment, State=#state{node_gets_total=NGT,
      {node_put_fsm_time_100, PutH},
-     {read_repairs_total, RRT}].
+     {read_repairs_total, RRT},
+     {coord_redirs_total, CRT}].
```
Please add to the module's doc, a description of what `coord_redirs_total` is.

beerriot commented on src/riak_object.erl:
```diff
@@ -101,29 +101,47 @@ ensure_robject(Obj = #r_object{}) -> Obj.
+    case vclock:equal(vclock(Obj1),vclock(Obj2)) of
+        false -> false;
+        true -> equal2(Obj1,Obj2)
+    end.
```
Why swap the short-circuiting boolean operations for more verbose case statements?

jonmeredith replied:
Rebase failure - code should match the current master.

beerriot commented on src/riak_object.erl:
```diff
@@ -137,9 +155,9 @@ reconcile(Objects, AllowMultiple) ->
     Contents = case AllowMultiple of
         false ->
-            [most_recent_content(AllContents)];
+            [hd(lists:sort(fun compare_content_dates/2, AllContents))];
```
Not that `hd(lists:sort(...` is hard to read, but wasn't `most_recent_content` more descriptive?

jonmeredith replied:
These are rebase failure - I'll try and work out the best way to fix.

beerriot commented:
I've noticed two differences between the AZ303 branch and current master so far.  I've put code in a gist (https://gist.github.com/445ece2b530c15a06008) to demonstrate each of:
- writes using vector clocks that are ancestors to the current vclock are no longer rejected, and instead cause siblings (`az303:ignore_old_vclocks/2`)
- allow_mult is no longer enforced at write time, so siblings are stored, even if only one value will be returned on read (`az303:enforce_allow_mult_on_put/2`)

Each function returns `pass` on master, but `{fail, Description, Object}` on az303 (apologies for the pass/fail semantics if this is an intended change).

beerriot commented:
It should be well-documented that no node should have `vnode_vclocks=true` before every node is running the az303 branch code.  Otherwise, objects may be written without vclocks.  To reproduce, start a cluster with one node running az303 and `vnode_vclocks=true`, then add 3 nodes running master to that cluster.  Write some objects, read them back, and you'll find vclocks of `[]`.

Commit:
Addressed review comments and fixed ec_eqc test for updated get_core.

Removed dead commented out code.
Added DeletedVclock arg to get_core
Renamed coord_idx_node to more meaningful coord_pl_entry.
Added code comment on requirement to upgrade all nodes to 1.0
Added code comment on coord_redits_total

Commit:
Propagated state element rename of coord_idx_node to put_fsm_eqc.

Commit:
Reverted accidental changes to riak_object for rebase.

My git foo doesn't extend to rewriting this kind of history.
Re-copied riak_object from master and re-applied changes.

jonmeredith commented:
Thx for the code examples.

The put behavior with out of date vclocks is by design - the idea is that if a user has said they want to put data we should try and do that even if it has been subsequently updated.  I do however need to add a way so we can do repl puts without changing the vclocks.

The allow_mult behavior should be enforced - I'll address that when I rebase after backend phase2 review is completed.

jonmeredith commented:
Superceded by https://github.com/basho/riak_kv/pull/166
